### PR TITLE
Perform only one TreeWalk to get files size when the number of files is bigger than a threshold (500)

### PR DESCRIPTION
**Pull request by dejan-milisavljevic-sonarsource**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -37,7 +37,7 @@ javadoc {
 dependencies {
     compileOnly 'com.google.code.findbugs:jsr305:3.0.2'
 
-    implementation 'org.eclipse.jgit:org.eclipse.jgit:6.4.0.202211300538-r'
+    implementation 'org.eclipse.jgit:org.eclipse.jgit:6.5.0.202303070854-r'
```
Comment (dbmeneses) [inline build.gradle:40]: This should be in a different commit with a dedicated ticket about the upgrade, so that it's easy to track it down if we end up having side effects due to this upgrade.

File: src/main/java/org/sonar/scm/git/blame/FileBlamer.java [hunk 2]
```diff
@@ -79,6 +91,22 @@ public void initialize(ObjectReader objectReader, GraphNode commit) {
     fileTreeComparator.initialize(objectReader);
   }
 
+  private void initializeForLargeFileSet(ObjectReader objectReader, GraphNode commit) {
+
+    Set<String> filesToCheck = commit.getAllFiles().stream().map(FileCandidate::getPath).collect(Collectors.toSet());
+    try {
+      Map<String, Integer> fileSizes = fileReader.getFileSizes(filesToCheck);
+      for (FileCandidate fileCandidate : commit.getAllFiles()) {
+        int fileSize = fileSizes.get(fileCandidate.getPath());
```
Comment (dbmeneses) [inline src/main/java/org/sonar/scm/git/blame/FileBlamer.java:100]: isn't there a risk that we didn't find the file during the tree walk and this ends up throwing a NPE?

File: build.gradle (hunk 1 above)
Comment (dejan-milisavljevic-sonarsource) [reply build.gradle:40]: I will revert that change. 
In SC this was overridden  by version 6.3.X and the performances were not good
We saw that in SQ, you are using 6.5.0.202303070854-r, so we did the same in SC and then blame was much faster
I will anyway revert it and create a new PR with the upgrade

File: src/main/java/org/sonar/scm/git/blame/FileBlamer.java (hunk 2 above)
Comment (dejan-milisavljevic-sonarsource) [reply src/main/java/org/sonar/scm/git/blame/FileBlamer.java:100]: Correct. I will add a check and throw and `IllegalStateException` as in the other alorithm using `loadText`

Comment (dejan-milisavljevic-sonarsource) [reply src/main/java/org/sonar/scm/git/blame/FileBlamer.java]: Done

File: build.gradle (hunk 1 above)
Comment (dejan-milisavljevic-sonarsource) [reply build.gradle]: Done

File: src/main/java/org/sonar/scm/git/blame/FileBlamer.java [hunk 3]
```diff
@@ -71,6 +74,15 @@ public FileBlamer(FileTreeComparator fileTreeComparator, DiffAlgorithm diffAlgor
    */
   public void initialize(ObjectReader objectReader, GraphNode commit) {
     this.objectReader = objectReader;
+    if (commit.getAllFiles().size() < NB_FILES_THRESHOLD_ONE_TREE_WALK) {
+      initializeForSmallFileSet(objectReader, commit);
+    } else {
+      // When we have more than NB_FILES_THRESHOLD_ONE_TREE_WALK we use this implementation which walks only once the entire tree
+      initializeForLargeFileSet(objectReader, commit);
```
[the remark this answers is not in the document]
Comment (dejan-milisavljevic-sonarsource) [reply src/main/java/org/sonar/scm/git/blame/FileBlamer.java:81]: I will run some more fine grained tests

**Comment by dejan-milisavljevic-sonarsource** [comment]
Performance comparison between 1 tree walk per file to blame, vs 1 tree walk
Java 11 MacBook Pro. Chip : Apple M2 Pro. Os version 13.4. Cores : 12 |   |  
-- | -- | --
  |   |  
Timing on file size computation. Average on 5 runs up to 1'000 files. 1 run only when more than 1'000 files. Done on a "new  repository" with no previous commits|   |  
  |   |  
NB files to blame | Multiple walks time in ms | 1 Walk time in ms
1 | 2 | 2
10 | 10 | 3
20 | 26 | 3
30 | 42 | 4
40 | 62 | 5
50 | 85 | 5
100 | 279 | 9
200 | 930 | 17
300 | 1,900 | 25
400 | 3,350 | 37
500 | 5,200 | 45
1,000 | 19,800 | 75
3,000 | 197,000 | 311
  |   |  
  |   |  
  |   |  
Timing on file size computation. Done on Typescript project, by keeping only X files to be blamed. All the other projects files are still in the repository |   |  
  |   |  
NB files to blame| Multiple walks time in ms | 1 Walk time in ms
1 | 5 | 1,200
10 | 15 | 1,220
20 | 30 | 1,222
30 | 50 | 1,230
40 | 55 | 1,235
50 | 1,700 | 1,250
100 | 9,700 | 1,270
200 | 25,500 | 1,300
300 | 41,500 | 1,320
400 | 57,000 | 1,380
500 | 72,300 | 1,400
1,000 | 152,000 | 1,500

File: src/main/java/org/sonar/scm/git/blame/FileBlamer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbmeneses) [reply src/main/java/org/sonar/scm/git/blame/FileBlamer.java:81]: thanks!

**Review by dbmeneses — APPROVED** [review]
